support for pickle protocol 4

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70409
* __->__ #70014
* #69793

Differential Revision: [D33264778](https://our.internmc.facebook.com/intern/diff/D33264778)

PaliC committed:
support for pickle protocol 4

[ghstack-poisoned]

suo commented on torch/package/_package_pickler.py:
```diff
@@ -94,9 +95,12 @@ def save_global(self, obj, name=None):
 
 
 def create_pickler(data_buf, importer):
+    protocol = 3
```
Oh, I didn't realize protocol 4 is supported since 3.4. In that case, we don't need versioning guards; pytorch itself only supports python 3.6+.

However, we should add a global flag that users can set for now so that they can disable protocol 4 if they are running into any weird behavior

PaliC replied:
Will do, but are there any examples of global flags lol

suo commented on torch/package/package_exporter.py:
```diff
@@ -585,24 +586,42 @@ def save_pickle(
+            field = None
             for opcode, arg, pos in pickletools.genops(data_value):
-                if opcode.name == "GLOBAL":  # a global reference
+                if sys.version_info >= (3, 4):
```
how do we know this is the only way that the arguments to STACK_GLOBAL get set up?

PaliC replied:
this is a versioning thing, we enforce pickle 4 from python 3.4 onwards.

suo reviewed (changes requested):
The implementation looks fine; it's not clear to me whether or not the opcodes you have here are exhaustive—is there something in the pickler or spec that makes us sure that these are the only ways STACK_GLOBAL will get invoked?

suo commented on torch/package/package_exporter.py:
```diff
@@ -585,24 +586,42 @@ def save_pickle(
+                        module = field
+                        field = None
+                    elif opcode.name == "STACK_GLOBAL":
+                        if module is None:
```
Why would module ever be None? Isn't the point of STACK_GLOBAL to load a module?

PaliC replied:
So this is specific to the set of 3 opcodes here. They grab some cached value which was likely already a global. Therefore, I just use None as a placeholder because we don't need to add the dependency again.

PaliC replied:
After taking a closer look at https://github.com/python/cpython/blob/main/Lib/pickle.py, I think this would work the vast majority of the time. However, it isn't obvious. to me how, but a valid order of opcodes could be to memoize the module name and then call with w/o a prior save global. Which this doesn't cover. I'll keep a dict for the the memoization in case this is an issue.

PaliC committed:
Update on "support for pickle protocol 4"




[ghstack-poisoned]

PaliC committed:
Update on "support for pickle protocol 4"




[ghstack-poisoned]

PaliC committed:
Update on "support for pickle protocol 4"




[ghstack-poisoned]

PaliC committed:
Update on "support for pickle protocol 4"




[ghstack-poisoned]

PaliC committed:
Update on "support for pickle protocol 4"




[ghstack-poisoned]

PaliC committed:
Update on "support for pickle protocol 4"




[ghstack-poisoned]

suo reviewed (approved):
lgtm, provided nothing breaks internally.

PaliC commented:
@sahancpal@gmail.com has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33264778).

PaliC committed:
Update on "support for pickle protocol 4"


Differential Revision: [D33264778](https://our.internmc.facebook.com/intern/diff/D33264778)

[ghstack-poisoned]

PaliC commented:
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33264778).
